GPU/Common/SplineCommon.cpp

```diff
@@ -361,16 +383,45 @@ static void _SplinePatchFullQuality(u8 *&dest, u16 *indices, int &count, const S
 	}
 }
 
+template <bool origNrm, bool origCol>
+static inline void SplinePatchFullQualityDispatch3(u8 *&dest, u16 *indices, int &count, const SplinePatchLocal &spatch, u32 origVertType, int quality, int maxVertices) {
+	bool origTc = (origVertType & GE_VTYPE_TC_MASK) != 0;
```

Comment:
Well, in this case, it doesn't exactly unroll it.  Rather, consider this:

```c++
void foo(bool abc) {
    double total = 0.0;
    for (int i = 0; i < 1000000; ++i) {
        if (abc) {
            total += 1.0;
        } else {
            total += 2.5;
        }
    }
}
```

A compiler is unlikely to unroll this, but it might flip things around (gcc -O3 probably):

```c++
void foo(bool abc) {
    double total = 0.0;
    if (abc) {
        for (int i = 0; i < 1000000; ++i) {
            total += 1.0;
        }
    } else {
        for (int i = 0; i < 1000000; ++i) {
            total += 2.5;
        }
    }
}
```

Which would of course avoid some branching on each iteration.  In this case, it's probably cheap - `abc` would be kept in a register, and probably there'd be little if any win.

But I was seeing the `if (abc) {` area as hot, even after SSE, so I factored it out.  Using the dispatch chains seemed like the simplest way to do this without duplicating code.

Loop unrolling is separate, it's when a loop is expanded, i.e.:

```c++
double quadruple(double v) {
    double total = 0.0;
    for (int i = 0; i < 4; ++i) {
        total += v;
    }
    return total;
}
```

Unrolled (gcc -O1 or whatever):

```c++
double quadruple(double v) {
    double total = 0.0;
    total += v;
    total += v;
    total += v;
    total += v;
    return total;
}
```

Both optimizations are applied automatically by compilers.  Neither MSVC or gcc (iirc) were flipping the checks for splines, which is why I used a template dispatch chain.

-[Unknown]